streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalProcessorContextImpl.java

```diff
@@ -59,11 +59,13 @@ public StateStore getStateStore(final String name) {
     }
 
     /**
-     * @throws UnsupportedOperationException on every invocation
+     * No-op. This should only be called on GlobalStateStore#flush and there should be no child nodes
      */
     @Override
     public <K, V> void forward(final K key, final V value, final To to) {
-        throw new UnsupportedOperationException("this should not happen: forward() not supported in global processor context.");
+        if (!currentNode().children().isEmpty()) {
+            throw new IllegalStateException("This method should only be called on 'GlobalStateStore.flush' that should not have any children.");
+        }
```

Comment:
@bbejeck @guozhangwang I was thinking about this once more, and for global tasks, we know that there are exactly two processors: the source processor and the "state maintainer" processor. The first one has exactly one child and it calls `forward()` from above. The second processor only call this `forward` overload on flush and should have zero children.

Does this make sense?

Reply:
There is a related JIRA about that but whether we'd keep it as is still open questions, I think we can make this assumption still atm but just bring it up FYI. 

https://issues.apache.org/jira/browse/KAFKA-7125